helper/helper.go

```diff
@@ -0,0 +1,81 @@
+package helper
+
+import (
+	"archive/tar"
+	"errors"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+)
+
+func GetApplication(applicationsPath, applicationUUID string) (string, string, error) {
+	commit, err := getCommit(applicationsPath, applicationUUID)
+	if err != nil {
+		log.Printf("Failed to get commit. Error: %s\r\n", err)
+		return "", "", err
+	}
+
+	basePath := filepath.Join(applicationsPath, applicationUUID, commit)
+	tarball := filepath.Join(basePath, "binary.tar")
+	application := filepath.Join(basePath, "application.zip")
+
+	_, err = os.Stat(application)
+	if err != nil {
+		err = untar(tarball, basePath)
+		if err != nil {
+			log.Printf("Failed to extract application. Error: %s\r\n", err)
+			return "", "", err
+		}
+	}
+
+	return application, commit, err
+}
+
+func untar(tarball, target string) error {
+	reader, err := os.Open(tarball)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tarReader := tar.NewReader(reader)
+	for header, err := tarReader.Next(); err != io.EOF; header, err = tarReader.Next() {
+		if err != nil {
+			return err
+		}
+
+		path := filepath.Join(target, header.Name)
+		info := header.FileInfo()
+		if info.IsDir() {
+			if err = os.MkdirAll(path, info.Mode()); err != nil {
+				return err
+			}
+			continue
+		}
+
+		file, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, info.Mode())
+		if err != nil {
+			return err
+		}
+		defer file.Close()
```

Comment:
I really don't like this. You're going to have open fd's for every single file you open which will only be closed on function exit. You'll easily exceed the kernel open file limit + cause the app to crash. `defer` is not always the correct approach, and _definitely_ is not here.

This is a critical issue so I'll ping - @josephroberts